### PR TITLE
Refactor DirectoryHelper

### PR DESCRIPTION
**Pull request by ThomasBarnekow**
Rename ```TryDeleteDirectory(string, int, int, int)``` to ```DeleteDirectory(string, int, int, int)``` because this method can still throw exceptions and is not fully in line with the try-parse pattern.

Introduce ```TryDeleteDirectory(string) : bool``` as a try-parse pattern-compliant option to delete directories recursively without throwing exceptions. It doesn't need an out parameter because the basic method it wraps does not return a value.

Remove ```DeleteSubdirectories(string)``` method because it is not used anywhere (housekeeping).

Use ```var``` keyword consistently across all methods.

Related to #980.

### CODE REVIEW

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs [hunk 1]
```diff
@@ -53,30 +59,28 @@ public static void DeleteDirectory(string directoryPath)
...
             NormalizeAttributes(directoryPath);
-            TryDeleteDirectory(directoryPath, maxAttempts: 5, initialTimeout: 16, timeoutFactor: 2);
+            DeleteDirectory(directoryPath, maxAttempts: 5, initialTimeout: 16, timeoutFactor: 2);
         }
 
         private static void NormalizeAttributes(string directoryPath)
         {
-            string[] files = Directory.GetFiles(directoryPath);
-            string[] dirs = Directory.GetDirectories(directoryPath);
+            var files = Directory.GetFiles(directoryPath);
```
Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: I may be picky, but I'd rather not hide types here.

**Comment by Therzok** [comment]
I don't like those var changes. Hiding types from stuff that's not clear from the right side of the expression is just not my style.

**Comment by ThomasBarnekow** [comment]
I'm not precious about these changes. Unless anybody else votes for keeping them, I could roll them back.

**Comment by ThomasBarnekow** [comment]
On the other hand, you could also say the variable names are suboptimal and, therefore, the type helps understand what these really are. For example, would you still disagree with the change if we used more descriptive variable names, like so?
```
var fileNames = Directory.GetFiles(directoryPath);
var directoryNames = Directory.GetDirectories(directoryPath);
```
From my point of view, it would make sense to rename these variables even if we roll back the previous change.

@nolltoken What do you think?

**Comment by nulltoken** [comment]
>I don't like those var changes. Hiding types from stuff that's not clear from the right side of the expression is just not my style.

:+1: We indeed tend to use `var` when the right side of the expression is explicit about the type being returned.

```
int i = 1;
var j = (long)i;
var things = new List<string>()
```

**Comment by ThomasBarnekow** [comment]
@Therzok and @nulltoken I've rolled back those changes but renamed the variables to be more explicit about what those really are.

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs [hunk 2]
```diff
@@ -15,6 +15,8 @@ public static class DirectoryHelper
             { "gitmodules", ".gitmodules" },
         };
 
+        private static readonly Type[] whitelist = { typeof(IOException), typeof(UnauthorizedAccessException) };
```
Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Why was this moved? :D

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs [hunk 3]
```diff
@@ -53,30 +59,28 @@ public static void DeleteDirectory(string directoryPath)
...
             File.SetAttributes(directoryPath, FileAttributes.Normal);
         }
 
-        private static readonly Type[] whitelist = { typeof(IOException), typeof(UnauthorizedAccessException) };
-
-        private static void TryDeleteDirectory(string directoryPath, int maxAttempts, int initialTimeout, int timeoutFactor)
+        private static void DeleteDirectory(string directoryPath, int maxAttempts, int initialTimeout, int timeoutFactor)
         {
-            for (int attempt = 1; attempt <= maxAttempts; attempt++)
+            for (var attempt = 1; attempt <= maxAttempts; attempt++)
```
Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: `int attempt`

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 2 above)
Comment (ThomasBarnekow) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: To have all fields at the top of the class. The other existing field also did not directly precede the (only) method in which it was used. In other words, consistency ...

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 3 above)
Comment (ThomasBarnekow) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: If you "use ```var``` when the right side of the expression is explicit about the type being returned", then ```var attempt = 1``` would be such a case. Wouldn't it?

Talking about ```var``` vs. explicit declarations, what would you say about the following?
```
var commits = repo.Commits.QueryBy(filter);
foreach (var reference in refsToRewrite.OrderBy(ReferenceDepth)) {}
```

Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: That's not `var attempt = (int)1`. It's not being explicit. See @nulltoken 's example above too.

I'd also have explicit typing there. How can I infer the generic type that's being used in the IEnumerable?

Unless the type doesn't _explicitly_ appear in the assignment instruction, do _not_ use `var`.

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 2 above)
Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Fine by me.

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 3 above)
Comment (ThomasBarnekow) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: @Therzok The two other ```var``` examples are from the ```HistoryRewriter``` class. Guess we'll have to "beautify" that as well then ... because you are absolutely right, you can't infer the generic type that's being used in the IEnumerable. The same would be true for the next real code example:
```
var refsToRewrite = repo.Refs.ReachableFrom(targetedCommits).ToList();
```

Looking at another example from ```HistoryRewriter```, would you be fine with 
```var success = false;```
as it stands or would you change that to
```bool success = false;```

I'd be fine with the ```var``` because ```false``` is a boolean literal and as per the C# language specification the type of a boolean literal is ```bool```. So I absolutely know that ```success``` is a ```bool```. Right?

By the same token (no pun intended ;-), other literals have well-defined types as well. For example, the character literal ```'c'``` is a ```char```, the string literal ```"Hello World"``` is a ```string```, and the real literals ```1.5``` (or ```1.5d```), ```1.5f```, and ```1.5m``` are a ```double``` (no type suffix, ```d```, or ```D```), ```float``` (type suffix ```f``` or ```F```), and ```decimal``` (type suffix ```m``` or ```M```), respectively. Therefore, the following should be fine:
```
var operand = '+';
var message = "Hello World!";
var amount = 1.5;
```
Would you agree?

An unsuffixed integer literal has the first of these types in which its value can be represented: ```int```, ```uint```, ```long```, ```ulong```. The integer literal ```1``` can clearly be represented as an ```int``` and will, therefore, definitely be an ```int```. Therefore, ```(int)``` in ```var amount = (int)1;``` is a redundant type cast (and ReSharper, for example, would tell you that). 

In other words, 
```
for (var attempt = 1; attempt <= maxAttempts; attempt++)
```
should be fine because the type, ``` int``` in case of the integer literal ```1``` on the right side (and also given the context of a ```for``` loop, where you typically have ```int``` loop variables), is absolutely clear from the right side of the expression.

Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Wow, this is going to take a while to comment. I'll comment inline.

> The two other var examples are from the HistoryRewriter class. Guess we'll have to "beautify" that as well then ... because you are absolutely right, you can't infer the generic type that's being used in the IEnumerable. The same would be true for the next real code example:
> 
> var refsToRewrite = repo.Refs.ReachableFrom(targetedCommits).ToList();
> Looking at another example from HistoryRewriter, would you be fine with 
> var success = false;
> as it stands or would you change that to
> bool success = false;

This is just a matter of preference, and I tend to swing on the explicit typing side of the wagon. Regardless of being keywords or not, I know the type's still there and an IDE can tell me. But having `var` everywhere hurts when reading and reviewing. Some things should help the eyes too, not let everything fall on the hands of tools - especially when we're talking about type inference.

Code can change, also. What if you decide to drop the initialization on the same line? You'd have to change it to an explicit type -> noise in diffs. If the type of a value assigned to a `var` variable changes, you will have unpredictable results. That's why I prefer explicit types.

> I'd be fine with the var because false is a boolean literal and as per the C# language specification the type of a boolean literal is bool. So I absolutely know that success is a bool. Right?

As said above, I _prefer_ the specification of types. Code changes.

> By the same token (no pun intended ;-), other literals have well-defined types as well. For example, the character literal 'c' is a char, the string literal "Hello World" is a string, and the real literals 1.5 (or 1.5d), 1.5f, and 1.5m are a double (no type suffix, d, or D), float (type suffix f or F), and decimal (type suffix m or M), respectively. Therefore, the following should be fine:
> 
> var operand = '+';
> var message = "Hello World!";

Let's also talk about some consistency. You can't use `var` in members. Let's only use it where it's nice on the eyes.

> var amount = 1.5;

I would argue here that you probably wanted a float and you'll get lots of warnings downhill about float and double mixing.

> Would you agree?
> 
> An unsuffixed integer literal has the first of these types in which its value can be represented: int, uint, long, ulong. The integer literal 1 can clearly be represented as an int and will, therefore, definitely be an int. Therefore, (int) in var amount = (int)1; is a redundant type cast (and ReSharper, for example, would tell you that).
> 

I only showed the casting as _an example_, not as a pattern. How would you specify _which_ of the values the variable is going to take? Is it going to be unsigned? Is it going to be signed? I know you're going to say - look at the value. But I'd have to scroll to the definition and _not_ look at the type, but at the value. What does the value convey me? Look at all the possible combinations. https://msdn.microsoft.com/en-us/library/aa664674%28v=vs.71%29.aspx

> In other words,
> 
> for (var attempt = 1; attempt <= maxAttempts; attempt++)
> should be fine because the type, int in case of the integer literal 1 on the right side (and also given the context of a for loop, where you typically have int loop variables), is absolutely clear from the right side of the expression.

Do I see the declaration of maxAttempts on the same line so I can infer it through skimming? It might be close, but when code grows, it goes out of context.

It's about having a growing readability standard on my behalf.

If @nulltoken agrees with your arguments, sure thing, I'll shut and stop complaining about nitpicks.

I'm just a big fan of code readability and planning fail-safe.

**Comment by ThomasBarnekow** [comment]
I'm also a very big fan of code readability. However, this is also driven by layout and redundancy, for example.

> This is just a matter of preference, and I tend to swing on the explicit typing side of the wagon.

Explicit is good, unless it is really redundant. Redundancy adds clutter. While that is not the case when you write ```int``` vs. ```var```, other situations benefit from the more "tabular" look and feel that using ```var``` brings about. For example:
```
int number = 1;
IEnumerable<string> strings = <something-that-obviously-returns-an-IEnumerable<string>>;
```
explicitly defines two variables. The following example is less explicit:
```
var number = 1;
var strings = <something-that-obviously-returns-an-IEnumerable<string>>;
```
It is clearly a matter of preference or judgement whether or not you would go with the second example. It is certainly nicer on my eyes and I would, therefore, prefer it as long as the righthand side is clearly enough (which is the judgement part) an ```int``` (which it is) or ```IEnumerable<string>``` (which depends on the concrete example).

> Code can change, also. What if you decide to drop the initialization on the same line? You'd have to change it to an explicit type -> noise in diffs.

Dropping the initialization on the same line will add noise no matter what. Changing ```int amount = 1;``` to ```int amount;``` adds the same noise as going from ```var amount = 1;``` to ```int amount;```.

> How would you specify which of the values the variable is going to take? Is it going to be unsigned? Is it going to be signed? I know you're going to say - look at the value.

There is a reason why the type of integer literals defaults to ```int``` (a signed integer) unless that integer literal can't be represented as an ```int```. It is because, in the vast majority of cases, a signed integer is appropriate and you want just that. Therefore, I'm totally fine with ```var amount = 1;``` for signed integers. If I wanted an unsigned integer, that would be an exceptional case in which I would definitely be explicit about my choice and write ```uint amount = 1;```. I would not even write ```var amount = 1u```, because it is so exceptional to me personally.

>> ```var amount = 1.5```

> I would argue here that you probably wanted a float and you'll get lots of warnings downhill about float and double mixing.

As for ```int``` being the default type for integer literals, there's a good reason why ```double``` is the default type for real literals. Most people in most cases seemingly don't want a ```float``` unless they have to use that type, e.g., because some API demands it. In case I had to use ```float``` I would again be explicit about it. Since ```double``` is the default, I'm fine with ```var``` when declaring and initializing a ```double``` at the same time, using a real literal (which would have to be suffixed to mean anything else than a double).

> But I'd have to scroll to the definition and not look at the type, but at the value.

Wouldn't you also have to scroll to the definition and look at the value "where it is nice on the eyes"?

Again, believe it or not, I'm not precious about ```var``` vs. ```int``` in this specific example. I actually changed it because ReSharper proposed it and I thought, "hey, that makes sense". I'd really be fine with rolling it back. So, @Therzok and @nulltoken, just tell me what you want me to do.

Personally, I now tend to use ```var``` more and more often and consistently. At the same time, I'm trying to use meaningful variable names so that I don't have to rely too much on the types. Wherever I must make assumptions on the characteristics of a collection, for example, I make sure the name is explicit about that (certainly not in the legacy prefix notation sense (e.g., ```lpstr```, though). This way, I don't even have to go back to the declaration, whether it's explicit or implicit, to understand what a variable represents.

**Comment by ThomasBarnekow** [comment]
Here you go ... I rolled it back and changed ```for (var attempt = 1; ...``` into ```for (int attempt = 1; ...```. Told you I was not precious about it ;-)

**Comment by ThomasBarnekow** [comment]
Travis said it was "unable to fetch some archives".

**Comment by nulltoken** [comment]
Build has been restarted

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs [hunk 4]
```diff
@@ -34,12 +36,16 @@ private static string Rename(string name)
             return toRename.ContainsKey(name) ? toRename[name] : name;
         }
 
-        public static void DeleteSubdirectories(string parentPath)
+        public static bool TryDeleteDirectory(string directoryPath)
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: As this method is no longer used, let's just drop it.

**Comment by nulltoken** [comment]
Other than the nitpick above, looks pretty nice to me. Thanks!

**Comment by ThomasBarnekow** [comment]
Will do.

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 4 above)
Comment (ThomasBarnekow) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Will do.

**Comment by ThomasBarnekow** [comment]
Deleted ```TryDeleteDirectory()```.

**Comment by nulltoken** [comment]
Another one in. Thanks! :sparkling_heart:

**Comment by ThomasBarnekow** [comment]
Thanks!
